### PR TITLE
SOC-1894 | Google login

### PR DESCRIPTION
**Pull request by kvas-damian**
https://wikia-inc.atlassian.net/browse/SOC-1894

Don't merge until entire google login and registration is done.

### CODE REVIEW

File: front/auth/app/externalauth/ExternalLogin.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		this.loginButton.classList.add('on');
+		this.loginButton.classList.add('disabled');
+	}
+
+	/**
+	 * @throws Error
+	 * @returns {void}
+	 */
+	onSuccessfulLogin() {
+		throw new Error('onSuccessfulLogin: Implement me!');
```
Comment (bkoval) [inline front/auth/app/externalauth/ExternalLogin.js]: That's a shockingly awesome message :100:

File: front/auth/app/google/GoogleLogin.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	/**
+	 * @returns {void}
+	 */
+	init() {
+		super.init();
+
+		window.gapi.load('auth2', () => {
+			const googleAuth = window.gapi.auth2.init({
```
Comment (bkoval) [inline front/auth/app/google/GoogleLogin.js]: Yeah, let's consider moving API initialization (lines 28 - 31) to the GoogleApi class and use callback injection for doing actions afterwards (lines 33 - 38).

File: front/auth/app/styles/module/_landing-page.scss [hunk 3]
```diff
@@ -39,18 +39,25 @@
...
 	margin-bottom: 15px;
 	opacity: 1;
 }
 
+.signup-provider-facebook {
+	background-color: #38528a;
+}
+
+.signup-provider-google {
+	background-color: #dd4b39;
```
Comment (bkoval) [inline front/auth/app/styles/module/_landing-page.scss]: Why don't we move both colors to variables? :)

Comment (kvas-damian) [inline front/auth/app/styles/module/_landing-page.scss]: Do we use them in other places? If not I don't see any reason why to do that.

File: server/app/facets/auth/register.js [hunk 4]
```diff
@@ -150,6 +187,8 @@ function getEmailRegistrationPage(request, reply) {
 export default function get(request, reply) {
 	if (request.query.method === 'facebook') {
```
Comment (rafalkalinski) [inline server/app/facets/auth/register.js]: I would like to see here some structure with methods and callbacks defined in it.

Comment (bkoval) [inline server/app/facets/auth/register.js]: What kind of structure? A lookup object?

Comment (rafalkalinski) [inline server/app/facets/auth/register.js]: yes :)

Comment (bkoval) [inline server/app/facets/auth/register.js]: OK, will do

File: front/auth/app/google/GoogleLogin.js (hunk 2 above)
Comment (kvas-damian) [inline front/auth/app/google/GoogleLogin.js]: Feel free to refactor it in your ticket ;P

File: front/auth/app/externalauth/ExternalLogin.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	/**
+	 * @returns {string}
+	 */
+	getRegistrationUrl() {
+		let search = window.location.search;
+
+		if (search.indexOf('?') !== -1) {
+			search += `&method=${this.getProviderName()}`;
```
Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: can this method... be moved to line 97 so it does not have to be twice here?

File: front/auth/app/google/GoogleLogin.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+			access_token: user.getAuthResponse().access_token
+		});
+	}
+
+	/**
+	 * TODO Remove this method and start using parent when SERVICES-1118 is resolved
+	 * @param {*} serviceRequestData
+	 * @returns {string}
+	 */
+	getInfoURL(serviceRequestData) {
```
Comment (hakubo) [inline front/auth/app/google/GoogleLogin.js]: getInfoURL({access_token}) {

File: server/app/facets/auth/register.js [hunk 7]
```diff
@@ -91,7 +92,41 @@ function getFacebookRegistrationPage(request, reply) {
...
+				headerText: 'auth:google-register.register-with-google',
+				heliosExternalRegistrationURL: authUtils.getHeliosUrl('/facebook/users'),
+				title: 'auth:google-register.register-with-google',
+				termsOfUseLink: 'http://www.wikia.com/Terms_of_Use',
+				footerCallout: 'auth:common.signin-callout',
+				footerHref: authUtils.getSignInUrl(request),
+				footerCalloutLink: 'auth:google-register.footer-callout-link',
+				bodyClasses: 'register-google-page',
+				pageType: 'register-google-page',
+				facebookAppId: localSettings.facebook.appId,
```
Comment (hakubo) [inline server/app/facets/auth/register.js]: do we need facebookAppId for getGoogleRegistrationPage?

Comment (kvas-damian) [inline server/app/facets/auth/register.js]: registration page is still in progress ;)

**Comment by rafalkalinski** [comment]
except linter failure, LGTM :)

File: front/auth/app/externalauth/ExternalLogin.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+		this.tracker = new AuthTracker('user-login-mobile', 'login');
+		this.loginButton = loginButton;
+		this.urlHelper = new UrlHelper();
+	}
+
+	/**
+	 * @returns {void}
+	 */
+	init() {
+		this.loginButton.addEventListener('click', this.login.bind(this));
```
Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: I don't see any teardown of this listener, are we worried about that?

Comment (kvas-damian) [inline front/auth/app/externalauth/ExternalLogin.js]: Do you think it matter for non single page apps?

Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: hence my question if we are worried about it :)

Since you are not - that is enough for me :)

Comment (bkoval) [inline front/auth/app/externalauth/ExternalLogin.js]: Login method deactivates a button, so there's no need for teardown here, especially that a result of this action can be either redirect or activating the button again :)

File: front/auth/app/externalauth/ExternalLogin.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	 */
+	login() {
+		this.deactivateButton();
+	}
+
+	/**
+	 * @returns {void}
+	 */
+	activateButton() {
+		this.loginButton.classList.remove('on');
```
Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: this.loginButton.classList.remove('on', 'disabled');

File: front/auth/app/externalauth/ExternalLogin.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+		this.loginButton.classList.remove('disabled');
+	}
+
+	/**
+	 * @returns {void}
+	 */
+	deactivateButton() {
+		this.loginButton.classList.add('on');
```
Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: this.loginButton.classList.add('on', 'disabled');

File: front/auth/app/externalauth/ExternalLogin.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+	onSuccessfulLogin() {
+		throw new Error('onSuccessfulLogin: Implement me!');
+	}
+
+	/**
+	 * @throws Error
+	 * @returns {string}
+	 */
+	getProviderName() {
+		throw new Error('getProviderName: Implement me!');
```
Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: what if in subclass I call
```js
super.getProviderName(...arguments)
```?
It'll throw which might be odd

File: front/auth/app/externalauth/ExternalLogin.js (hunk 9 above)
Comment (kvas-damian) [inline front/auth/app/externalauth/ExternalLogin.js]: It's not supported on IE.

File: front/auth/app/externalauth/ExternalLogin.js (hunk 11 above)
Comment (kvas-damian) [inline front/auth/app/externalauth/ExternalLogin.js]: Do you think we should just remove this method here?

Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: dunno, would like to chat about that. Don't have strong opinion on that for now.
I still have to wrap my head around if using class in JS should be encouraged :)

File: front/auth/app/externalauth/ExternalLogin.js (hunk 9 above)
Comment (hakubo) [inline front/auth/app/externalauth/ExternalLogin.js]: gotcha! did not know about that. caniuse to the rescue!

**Comment by kvas-damian** [comment]
Will open again when dependencies are ready.

**Pull request closed without merging**